### PR TITLE
feat(transform): Add hash string conversion

### PR DESCRIPTION
**Pull request by MisterMX**
### Description of your changes

This adds three new string conversion function to patch transforms, `ToSha1`, `ToSha256`, `ToSha512`, that generate a hash from the JSON representation of the input value.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

`composition.yaml`

```yaml
---
apiVersion: apiextensions.crossplane.io/v1
kind: Composition
metadata:
  name: test-string-tojson
spec:
  compositeTypeRef:
    apiVersion: example.org/v1alpha1
    kind: XSecret
  resources:
    - name: secret
      base:
        apiVersion: secretsmanager.aws.crossplane.io/v1beta1
        kind: Secret
        metadata:
          name: # patched
          annotations:
            hash: # patched
        spec:
          forProvider:
            region: eu-central-1
            stringSecretRef:
              name: # patched
              namespace: # patched
              key: # patched
          providerConfigRef:
            name: example
      patches:
        # Policy
        - fromFieldPath: "spec.parameters"
          toFieldPath: "metadata.annotations[hash]"
          transforms:
            - type: string
              string:
                type: Convert
                convert: ToSha1
        # SecretRef
        - fromFieldPath: "spec.parameters.secretRef.name"
          toFieldPath: "spec.forProvider.stringSecretRef.name"
        - fromFieldPath: "spec.parameters.secretRef.key"
          toFieldPath: "spec.forProvider.stringSecretRef.key"
        - fromFieldPath: "metadata.labels[crossplane.io/claim-namespace]"
          toFieldPath: "spec.forProvider.stringSecretRef.namespace"
```

`xrd.yaml`

```yaml
---
apiVersion: apiextensions.crossplane.io/v1
kind: CompositeResourceDefinition
metadata:
  name: xsecrets.example.org
spec:
  group: example.org
  claimNames:
    kind: Secret
    plural: secrets
  names:
    kind: XSecret
    plural: xsecrets
  versions:
    - name: v1alpha1
      served: true
      referenceable: true
      schema:
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                parameters:
                  type: object
                  properties:
                    secretRef:
                      type: object
                      description: The Kubernetes Secret whose data will be sent as string to AWS.
                        The secret needs to exist in the same namespace as this claim.
                      properties:
                        key:
                          type: string
                          description: Optional key whose value should be uploaded. If not set, the
                            whole map in the Secret data will be used.
                        name:
                          type: string
                          description: Name of the Kubernetes Secret resource in the same namespace
                            as this claim.
                      required:
                        - name
              required:
                - parameters
```

`xrc.yaml`

```yaml
---
apiVersion: v1
kind: Secret
metadata:
  name: example-secret
  namespace: default
type: Opaque
stringData:
  password: passw0rd
---
apiVersion: example.org/v1alpha1
kind: Secret
metadata:
  name: example-secret
  namespace: default
spec:
  parameters:
    secretRef:
      name: example-secret
```

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: internal/controller/apiextensions/composite/transforms.go [hunk 1]
```diff
@@ -258,11 +263,29 @@ func stringConvertTransform(input any, t *v1.StringConversionType) (any, error)
...
+	case v1.StringConversionTypeToSHA512:
+		hash, err := stringGenerateHash(input, sha512.Sum512)
+		return hex.EncodeToString(hash[:]), errors.Wrap(err, errHash)
 	default:
 		return nil, errors.Errorf(errStringConvertTypeFailed, *t)
 	}
 }
 
+func stringGenerateHash[THash any](input any, hashFunc func([]byte) THash) (THash, error) {
+	inputJSON, err := json.Marshal(input)
```
Comment (turkenh) [inline internal/controller/apiextensions/composite/transforms.go:281]: Is this the first usage of generics in Crossplane 🤔

File: docs/reference/composition.md [hunk 2]
```diff
@@ -734,6 +734,12 @@ Currently only `multiply` is supported.
      type: Convert
      convert: ToJson
 
+# The output will be the hash of the JSON representation of the 'from' field.
+- type: string
+  string:
+     type: Convert
+     convert: ToSha1 # alternatives: 'ToSha256' or 'ToSha512'
```
Comment (turkenh) [inline docs/reference/composition.md:741]: Please update the documentation under https://github.com/crossplane/docs and share the PR link here.

File: internal/controller/apiextensions/composite/transforms.go (hunk 1 above)
Comment (MisterMX) [reply internal/controller/apiextensions/composite/transforms.go:281]: Its a feature of the newer Go version. Is that a problem? Its only a very small function.

File: docs/reference/composition.md (hunk 2 above)
Comment (MisterMX) [reply docs/reference/composition.md:741]: Here you go: https://github.com/crossplane/crossplane/pull/3498

Comment (turkenh) [reply docs/reference/composition.md]: The link points to the PR itself?

Comment (turkenh) [reply docs/reference/composition.md]: Nm, I just found, already merged https://github.com/crossplane/docs/pull/293

**Review by turkenh — APPROVED** [review]
Thanks @MisterMX!

Looks like we need a rebase here too.

**Pull request merged**